File: Dijkstra/skip_list.py
import random
import numpy
#import fastrand


class SkipNode:   
    def __init__(self, height = 0, value = None, info = None):
        self.next = [None]*height #list of pointers
        self.value = value #value for list sorting
        self.info = info #other information


class SkipList:
    def __init__(self,min_switch=0): #initial list
        self.head = SkipNode() #first node, no information
        self.len = 0 #list size
        self.maxHeight = 0 #max height
        self.switch=-1 #originally max priority queue
        if min_switch: #switch: change to min priority queue
            self.switch=1
        
    def __len__(self): #get the list size
        return self.len
    
    def find(self, value, update = None): #get node with the given value
        if update == None:
            update = self.updateList(value) #path to the given value
        if len(update) > 0:
            candidate = update[0].next[0] #bottom node of the entire route
            if candidate != None:
                if candidate.value == value:
                    return candidate
        return None #cannot find node with this value
   
    '''
    def find_by_name(self, value, node_info, update = None): #get node with the given value
        if update == None:
            update = self.updateList(value) #path to the given value
        if len(update) > 0:
            candidate = update[0].next[0] #bottom node of the entire route
            if candidate != None:
                print("candinfo: " + str(candidate.info))
                print("candval: " + str(candidate.value))
                if candidate.value == value and candidate.info == node_info:
                    return candidate
    '''

    def print_list(self):
        for i in range(self.maxHeight-1, -1, -1):
            print("level:",i, end = " ")
            curr = self.head.next[i]
            while curr != None:
                print(curr.info, end = "->")
                curr = curr.next[i]
            print()


    def find_by_name(self, value, node_info, update = None): #get node with the given value
        if update == None:
            update = self.updateListInfo(value, node_info) #path to the given value
        if len(update) > 0:
            candidate = update[0].next[0] #bottom node of the entire route
            #found = None
            if candidate != None and candidate.value == value and candidate.info == node_info:
                return candidate
            return None


    def __contains__(self, value): #check whether a value in this list
        return self.find(value) != None

    def flip_coin(self): #determine the height of the new node
        count = 1
        while random.randint(0,1)!=0: #if toll head, have one more hierarchy
        #while numpy.random.randint(0,2)!=0: #if toll head, have one more hierarchy
        #while fastrand.pcg32bounded(2) != 0:
            count+=1
        return count

    def updateListInfo(self, value:float, info:tuple): #get a route to the node
        update = [None]*self.maxHeight #save the route
        x = self.head #from the left headnode
        for i in reversed(range(self.maxHeight)): #from top to bottom
            #while x.next[i] != None and x.next[i].value < value: #next_value>=given_value => walk down
            #while x.next[i] != None and x.next[i].value < value and x.next[i].info < info: #next_value>=given_value => walk down
            while x.next[i] != None and x.next[i].value < value: 
                x = x.next[i] 
            while x.next[i] != None and x.next[i].value == value and x.next[i].info < info: #next_value>=given_value => walk down
                x = x.next[i] #go to next value
            update[i] = x #save the route
            # print("Update list info [level",i,"]:", x.info, info, value)
        return update
        
    def updateList(self, value:float): #get a route to the node
        update = [None]*self.maxHeight #save the route
        x = self.head #from the left headnode
        for i in reversed(range(self.maxHeight)): #from top to bottom
            while x.next[i] != None and x.next[i].value < value: 
                x = x.next[i] 
            update[i] = x #save the route
        return update

    def add(self,input_tuple, debug=False):
        #self.switch determines min_heap(1) or max_heap(-1)
        value=input_tuple[0]*self.switch
        info=input_tuple[1]
        if debug:
            print("adding " + str(info) + ": " + str(value))
        #flip coin to get the height of the new node
        node = SkipNode(self.flip_coin(), value , info)
        #update skip list if has new highest: maxHeight, head_node
        self.maxHeight = max(self.maxHeight, len(node.next))
        if len(self.head.next) < self.maxHeight:
            self.head.next.extend([None]*(self.maxHeight- len(self.head.next)))
        #find route to the given value
        update = self.updateListInfo(value, info)
        #update every node in the given route
        for i in range(len(node.next)):
            #every rank point to the next
            node.next[i] = update[i].next[i]
            #update the left node
            update[i].next[i] = node
        self.len += 1

    def remove(self, value):
        #find the route to the given value
        update = self.updateList(value)
        #whether the value exist, the route can accelerate the find progress
        x = self.find(value, update)
        #if find the value
        if x != None:
            for i in reversed(range(len(x.next))):
                #update every rank, change the left node pointer
                update[i].next[i] = x.next[i]
                #check whether this node is the only node in this rank
                if self.head.next[i] == None:
                    #if it is, decrease the height of the skiplist
                    self.maxHeight -= 1
            #change list size
            self.len -= 1
        # x.info means removed, 0 means cannot find the value
            return x.info
        return 0
    
    def remove_by_name(self, value, node_info, update=None):
        #find the route to the given value
        value = value*self.switch 
        #print("removing by " + str(node_info))
        #print("removing by " + str(value))
        if update == None:
            update = self.updateListInfo(value, node_info)
        #whether the value exist, the route can accelerate the find progress
        # print(update[0].info)
        x = self.find_by_name(value, node_info, update)
        #if find the value
        if x != None:
            for i in reversed(range(len(x.next))):
                update[i].next[i] = x.next[i]
                #check whether this node is the only node in this rank
                if self.head.next[i] == None:
                    #print("Decreasing height")
                    #if it is, decrease the height of the skiplist
                    self.maxHeight -= 1
            #change list size
            self.len -= 1
        #1 means removed, 0 means cannot find the value
            return 1
        return 0


    def __str__(self):
        #return the list as a string
        result=[]
        x = self.head
        #append every node in the lowest rank
        while x.next[0] != None:
            x = x.next[0]
            result.append((x.value*self.switch,x.info))
        return str(result)

    def __getitem__(self,index):
        #return value by index
        #very slow, like find items by index in linked list
        #check index
        assert(index>=0)
        assert(index<self.len)
        #iteration find node
        x = self.head
        i = -1
        while True:
            if i==index:
                return (x.value*self.switch,x.info)
            if x.next[0] == None:
                break
            i=i+1
            x = x.next[0]
        return None

    def updateListEnd(self, value:float): #get a route to the node
        update = [None]*self.maxHeight #save the route
        x = self.head #from the left headnode
        for i in reversed(range(self.maxHeight)): #from top to bottom
            while x.next[i] != None and x.next[i].value <= value:
                x = x.next[i]
            update[i] = x #save the route
        return update

    def findEnd(self, value, update=None):
        if update == None:
            update = self.updateListEnd(value) #path to the given value
        if len(update) > 0:
            candidate = update[0]
            if candidate != None:
                if candidate.value == value:
                    return candidate
        return None #cannot find node with this value

    def removeRand(self, value, start=None, end=None):
        # find the start nodes and the end nodes that have such value
        if start == None:
            start = self.updateList(value)[0].next[0]
        if end == None:
            end = self.updateListEnd(value)[0]
        # print("nodes range for value", value, "is [", start.info, end.info, "]")
        # assert start.value == end.value, "start and end have different value!"+ str(start.value) + str(end.value)
        # random choose one info
        g1_info = random.randint(min(start.info[0], end.info[0]), max(start.info[0], end.info[0]))
        g2_info = random.randint(min(start.info[1], end.info[1]), max(start.info[1], end.info[1]))
        # print("Random node info:", (g1_info,g2_info))
        # get the nearest path to that node
        update = self.updateListInfo(value, (g1_info, g2_info))
        # TODO: low probability to choose the last one?
        # cand1 = update[0]
        # cand2 = update[0].next[0]
        # if cand1 and cand2 and g1_info + g2_info - cand1.info[1] -cand2.info[2] < cand1.info[1]

        cand1 = update[0] if update[0] and update[0].info != None else update[0].next[0]
        # print("candidate for removeRand", cand1.info, cand1.value)
        self.remove_by_name(cand1.value*self.switch, cand1.info)
        return (cand1.value, cand1.info)

    def pop(self, domain=0.1, fast=True):
        if fast:
            # random value, but the first one with that value
            return self.popfast(domain)
        else:
            return self.popUniform(domain)


    def popUniform(self, domain=0.1):
        # check not empty
        if self.__len__() == 0:
            raise IndexError()
        # get a random value in the given random value range
        rand_value = random.uniform(self.head.next[0].value, self.head.next[0].value + domain)
        cand1 = self.updateList(rand_value)[0]
        cand2 = self.updateListEnd(rand_value)[0]
        # print(cand1.value, cand2.value)
        if cand1 and cand2 and cand1.next[0] and cand1.next[0].value == cand2.value:
            ret = self.removeRand(cand1.value, start=cand1, end=cand2)
        elif cand2 and abs(cand2.value-rand_value) < abs(cand1.value-rand_value):
            ret = self.removeRand(cand2.value, start=cand2, end=None)
        else:
            ret = self.removeRand(cand1.value, start=None, end=cand1)
        return ret


    def popfast(self, domain=0.1):
        '''Pop out the first node with a random value'''
        #check not empty
        if self.__len__()==0:
            raise IndexError()
        #get a random value in the given random value range
        #rand_value=random.uniform(self.head.next[0].value,self.head.next[0].value+domain)
        rand_value=random.uniform(self.head.next[0].value, self.head.next[0].value+domain)
        #if this value exist, directly remove it
        if self.__contains__(rand_value):
            info = self.remove(rand_value)
            return (rand_value * self.switch, info)
        #if this value not in list, remove the nearest one
        else:
            update = self.updateList(rand_value)
            #get the left one & right one, compare them
            candidate1 = update[0]
            candidate2 = update[0].next[0]
            #if the right one not exist, return the left one
            if candidate2==None:
                info = self.remove(candidate1.value)
                return (candidate1.value*self.switch,info)
            #otherwise compare them
            if abs(candidate1.value-rand_value)<=abs(candidate2.value-rand_value):
                info = self.remove(candidate1.value)
                return (candidate1.value*self.switch,info)
            if abs(candidate2.value-rand_value)<abs(candidate1.value-rand_value):
                info = self.remove(candidate2.value)
                return (candidate2.value*self.switch,info)
            return None

    def opop(self, domain=0.1):
        '''Reservoir Sampling: slow!'''
        #that's the new one
        if self.__len__()==0:
            raise IndexError()
        tail=self.head.next[0].value+domain
        temp=self.head.next[0]
        choice=(temp.value,temp.info)
        n=0
        while temp and temp.value<tail :
            if temp.next[0]==None:
                break
            temp=temp.next[0]
            n+=1
            
            #j=random.randint(0,n)
            j=numpy.random.randint(0,n+1)
            if j==0:
                choice=(temp.value,temp.info)
        self.remove(choice[0])
        return (choice[0]*self.switch,choice[1])

File: Dijkstra/SkipList.c

// Node definition
struct SkipListNode
{
    SkipListNode[] next;
    size_t height;
    float value;
    void** info;
}

// Skip List definition
struct SkipList
{
    SkipListNode* head;
    size_t len;
    size_t maxHeight;
    int switch;
}

// Node initializer
SkipListNode* node_init(size_t height = 0, float value = NULL, void** info= NULL)
{
    struct SkipListNode self;
    self.next = (SkipListNode*)calloc(height, sizeof(SkipListNode*));
    self.height = height;
    self.value = value;
    self.info = info;
    return &self;
}

// Skip List Initializer
SkipList* list_init(bool min)
{
    struct SkipList self;
    self.head = node_init();
    self.len = 0;
    self.maxHeight = 0;
    self.switch = min ? -1 : 1;
    return &self;
}

// coin flip to determine height of new node 
// return height of new node
// TODO: better random generator (?)
int flip_coin()
{
    int height = 1;
    srand(time(NULL));
    while(rand()%2 != 0)
    {
        height++;
    }
    return height;
}

// find route to node, by value, through all levels 
// return array of last nodes at each level such that node.value < value at each level
SkipListNode** updateList(SkipList* self, float value)
{
    SkipListNode* path[];
    SkipListNode* x = self->head;
    for(int i = self->maxHeight-1; i>-1; i--)
    {
        while(x->next[i] != NULL && x->next[i]->value < value)
        {
            x = x->next[i];
        }
        path[i] = x;
    }
    return path;
}

// find a node by value 
// uses updateList (<) to find path
// return pointer to node
SkipListNode* find(SkipList* self, float value, SkipListNode** update = NULL)
{
    if(update == NULL)
    {
        update = updateList(self, value);
    }
    candidate = update[0]->next[0];
    if(candidate != NULL)
    {
        if(candidate->value == value)
        {
            return candidate;
        }
    }
    return NULL;
}

// check if value is in list
// use find()
// return true if found, false otherwise
bool __contains__(SkipList* self, float value)
{
    return (bool)(self.find(value) != NULL);
}


// insert a node with the given value and info
// uses updateListInfo
// return 0 on success, -1 on failure
int add(SkipList* self, float value, void** info)
{
    SkipList* node = node_init(flip_coin(), value, info);
    self->maxHeight = max(self->maxHeight, node->height);
    // if self->head->height < self->maxHeight, extend the head
    update = self.updateListInfo(value);
    for(int i = 0; i<node->height; i++)
    {
        node->next[i] = update[i]->next[i];
        update[i]->next[i] = node;
    }
    self->len++;
    return 0;
}

// remove node with value from the list
// uses updateList
// return 0 on failure, info on success
void** remove(self, value):
{
    x = find(self, value, update);
    if(x != NULL)
    {
        for(int i = x->height-1; i > -1; i--)
        {
            update[i]->next[i] = x->next[i];
            if(self->head->next[i] == NULL)
            {
                self->maxHeight--;
            }
        }
        self->len--;
        return x->info;
    }
    return 0;
}


// find route to node, by value, through all levels
// return array with last nodes at each level such that node.value <= value at each level
SkipListNode** updateListEnd(SkipList* self, float value)
{
    SkipListNode* path[];
    SkipListNode* x = self->head;
    for(int i = self->maxHeight-1; i>-1; i--)
    {
        while(x->next[i] != NULL && x->next[i]->value <= value)
        {
            x = x->next[i];
        }
        path[i] = x;
    }
    return path;
}

// find a node by value 
// uses updateListEnd (<=) to find path
// return pointer to node
SkipListNode* findEnd(SkipList* self, float value, SkipListNode** update=NULL)
{
    if(update == NULL)
    {
        update = updateListEnd(self, value);
    }
    candidate = update[0]->next[0];
    if(candidate != NULL)
    {
        if(candidate->value == value)
        {
            return candidate;
        }
    }
    return NULL;
}


// find route to node, by value, through all levels 
// return array with last node s.t. node.value<value and node.info<info at each level
SkipListNode** updateListInfo(SkipList* self, float value, void** tuple)
{
    SkipListNode* path[];
    SkipListNode* x = self->head;
    for(int i = self->maxHeight-1; i>-1; i--)
    {
        while(x->next[i] != NULL && x->next[i]->value < value)
        {
            x = x->next[i];
        }
        while(x->next[i] != NULL && x->next[i]->value == value && x->next->info<info)
        {
            x = x->next[i];
        }
        path[i] = x;
    }
    return path;
}

// remove node with value and info from the list
// return 1 on success, 0 if node not found
int remove_by_name(SkipList* self, float value, void** node_info, SkipListNode** update=None):
{
    float value = value*(self->switch);
    if(update == NULL)
    {
        update = updateListInfo(self, value, node_info);
    }
    SkipListNode* x = find_by_name(self, value, node_info, update);
    if(x != NULL)
    {
        for(int i = x->next->height-1; i > -1; i--)
        {
            update[i]->next[i] = x->next[i];
            if(self->head->next[i] == NULL)
            {
                self->maxHeight--;
            }
        }
        self->len--;
        return 1;
    }
    return 0;
}

// find a node by info 
// (return pointer to node)
SkipListNode* find_by_name(SkipList* self, float value, void** node_info, SkipListNode** update = NULL)
{
    if(update == NULL)
    {
        update = updateListInfo(self, value, node_info);
    }
    candidate = update[0]->next[0];
    if(candidate != NULL && candidate->value == value && candidate->info == node_info)
    {
        return candidate;
    }
    return NULL;
}

// print out full list
// return nothing
void print_list(SkipList* self)
{
    for(int i = self->maxHeight-1, i>-1, i--)
    {
        printf("level: %d", i);
        curr = self->head->next[i];
        while(curr != NULL)
        {
            printf("%d -> ", curr->value);
            // printf(curr->info);
            curr = curr->next[i];
        }
        printf("\n");
    }
}

// return node at index in list
// NULL if not found
SkipListNode* __getitem__(SkipList* self, int index):
{
    assert(index>=0);
    assert(index<self->len);
    SkipListNode* x = self->head;
    int i = -1;
    while(true)
    {
        if(i==index)
        {
            return x;
        }
        if(x->next[0] == NULL)
        {
            break;
        }
        i++;
        x = x->next[0];
    }
    return NULL;
}

// remove random node with given value
// uses updateList and updateListEnd to find upper and lower bounds
// return value and info of removed node 
// (just info for now until I figure out a good way to do this)
void** removeRand(SkipList* self, float value, SkipListNode** start=NULL, SkipListNode** end=NULL)
{
    if(start == NULL)
    {
        start = updateList(self, value)[0]->next[0];
    }
    if(end == NULL)
    {
        end = updateListEnd(self, value)[0]
    }
    srand(time(NULL));
    int max0 = max(start->info[0], end->info[0]);
    int min0 = min(start->info[0],end->info[0]);
    int diff0 = max0-min0;
    int max1 = max(start->info[1], end->info[1]);
    int min1 = min(start->info[1],end->info[1]);
    int diff1 = max1 - min1;
    int g1_info = min0 + rand()%(diff0);
    int g2_info = min1 + rand()%(diff1);
    int[] g_info = {g1_info, g2_info};

    SkipListNode** update = updateListInfo(self, value, g_info);

    SkipListNode* = (update[0] && update[0]->info != NULL) ? update[0] : update[0]->next[0];

    remove_by_name(self, (self->switch)*(cand1->value), cand1->info);
    return cand1->info;
}


// if fast, popfast(domain)
// else, popUniform(domain)
// return whatever the relevant function returns
// (going to make other pop functions return info for now)
void** pop(self, domain=0.1, fast=True):
{
    if(fast)
    {
        return self.popfast(domain);
    }
    else
    {
        return self.popUniform(domain);
    }
}

// remove a random node in the range (node[0].value, node[0].value+domain)
// use removeRand once the random value has been selected
// return the same as removeRand
void** popUniform(SkipList* self, float domain=0.1):
{
    assert(self->len > 0);
    srand(time(NULL));
    float rand_value = self->head->next[0]->value+ domain*(float)rand()/(float)(RAND_MAX);
    SkipListNode* cand1 = updateList(self,rand_value)[0];
    SkipListNode* cand2 = updateListEnd(self,rand_value)[0];

    if (cand1 && cand2 && cand1->next[0] && cand1->next[0]->value == cand2->value)
    {
        void** ret = removeRand(self, cand1->value, start = cand1, end= cand2);
    }
    else
    {
        if(cand2 && abs(cand2->value-rand_value) < abs(cand1->value-rand_value))
        {
            void** ret = removeRand(self, cand2->value, start = cand2, end=None);
        }
        else
        {
            void** ret = self.removeRand(cand1.value, start = None, end = cand1);
        }
    }
    return ret;
}

// pop first one that works for a randomly generated value
void** popfast(SkipList* self, float domain=0.1):
{
    assert(self->len > 0);
    srand(time(NULL));
    float rand_value = self->head->next[0]->value+ domain*(float)rand()/(float)(RAND_MAX);
    if(__contains__(self, rand_value))
    {
        void** info = remove(self, rand_value);
        return info;
    }
    else
    {
        SkipListNode** update = updateList(self, rand_value);
        SkipList* candidate1 = update[0];
        SkipList* candidate2 = update[0]->next[0];

        if(candidate2 == NULL)
        {
            void** info = remove(self, candidate1->value);
            return info;
        }
        if(abs(candidate1->value-rand_value)<=abs(candidate2->value-rand_value))
        {
            void** info = remove(self, candidate1->value);
            return info;
        }
        if(abs(candidate2.value-rand_value)<abs(candidate1.value-rand_value))
        {
            void** info = remove(self, candidate2->value);
            return info;
        }
        return NULL;
    }
}

// another removal algorithm
void** opop(SkipList* self, float domain=0.1):
{
    assert(self->len > 0);
    float tail = self->head->next[0]->value + domain;
    SkipListNode* temp = self->head->next[0];
    float choice_value = temp->value;
    void** choice_info = temp->info;

    int n = 0;
    
    srand(time(NULL));
    while(temp && temp->value < tail)
    {
        if(temp->next[0] == NULL)
        {
            break;
        }
        temp = temp->next[0];
        n++;

        int j = rand()%(n+2);
        if(j==0)
        {
            choice_value = temp->value;
            choice_info = temp->info;
        }
    }
    remove(self, choice_value);
    return choice_info;
}

// TODO: figure out a good way to do this in C
// return list as a string
// treat as regular linked list:
// only worry about bottom level
char* __str__(SkipList* self):
{
    return NULL;
}


int main()
{
    return 0;
}